src/Microsoft.NetCore.Analyzers/UnitTests/Security/DoNotUseInsecureDeserializerBinaryFormatterWithoutBinderTests.cs

```diff
@@ -0,0 +1,602 @@
+            return formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}",
+            GetCSharpResultAt(12, 20, BinderNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+        }
+
+        // Ideally, we'd detect that formatter.Binder is always null.
+        [Fact]
+        public void DeserializeWithInstanceField_Diagnostic_NotIdeal()
+        {
+            VerifyCSharp(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        BinaryFormatter formatter = new BinaryFormatter();
+
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            return this.formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}",
+            GetCSharpResultAt(13, 20, BinderMaybeNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+        }
+
+        [Fact(Skip = "Ideally, we'd detect that this.formatter.Binder is set.")]
+        public void DeserializeWithInstanceField_NoDiagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        BinaryFormatter formatter = new BinaryFormatter() { Binder = new MyBinder() };
+
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            return this.formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public void Deserialize_BinderMaybeSet_Diagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+            if (Environment.MachineName.StartsWith(""a""))
+            {
+                formatter.Binder = new MyBinder();
+            }
+
+            return formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}",
+            GetCSharpResultAt(18, 20, BinderMaybeNotSetRule, "object BinaryFormatter.Deserialize(Stream serializationStream)"));
+        }
+
+        [Fact]
+        public void Deserialize_BinderSet_NoDiagnostic()
+        {
+            VerifyCSharpWithMyBinderDefined(@"
+using System;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Blah
+{
+    public class Program
+    {
+        public object BfUnsafeDeserialize(byte[] bytes)
+        {
+            BinaryFormatter formatter = new BinaryFormatter();
+            formatter.Binder = new MyBinder();
+            return formatter.Deserialize(new MemoryStream(bytes));
+        }
+    }
+}");
+        }
+
+        [Fact]
+        public void TwoSettersOneBinderOnFirst_Diagnostic()
```

Comment:
>TwoSettersOneBinderOnFirst_Diagnostic [](start = 20, length = 37)

Do you have a no-diagnostic test where binder is a local variable that is assigned different non-null values in both if/else parts and then assigned to `formatter.Binder` prior to invoking `Deserialize`?

Reply:
Also consider adding a test where binder is a parameter, and you have `binder = binder ?? new MyBinder()` and then create `new BinaryFormatter() { Binder = binder }`

---
In reply to: [224095152](https://github.com/dotnet/roslyn-analyzers/pull/1838#discussion_r224095152) [](ancestors = 224095152)

Reply:
Adding `BinderVariableSetInAllBranches_NoDiagnostic()`, `BinderParameter_NoDiagnostic()`, and `BinderNotNullInsideIf_NoDiagnostic()`